src/hardware/gameblaster.cpp

```diff
@@ -16,105 +17,178 @@
+	static frame_t input = {};
+	static int16_t *buffer[] = {&input[0], &input[1]};
+	static device_sound_interface::sound_stream stream;
+
+	frame_t output = {};
+	for (const auto &d : devices) {
+		d->sound_stream_update(stream, 0, buffer, 1);
+		output[0] += input[0];
+		output[1] += input[1];
 	}
+	return output;
 }
 
-static void CMS_CallBack(Bitu len) {
-	enum {
-		BUFFER_SIZE = 2048
-	};
+void GameBlaster::RenderForMs(const double duration_ms)
+{
+	auto render_count = iround(duration_ms * frame_rate_per_ms);
+	while (render_count-- > 0)
+		fifo.emplace(RenderOnce());
+}
 
-	if ( len > BUFFER_SIZE )
-		return;
+void GameBlaster::RenderUpToNow()
+{
+	const auto now = PIC_FullIndex();
+	if (!is_enabled) {
+		assert(channel);
+		assert(fifo.empty());
+		channel->Enable(true);
+		is_enabled = true;
+	} else {
+		RenderForMs(now - last_render_time);
+	}
+	last_render_time = now;
+	unwritten_for_ms = 0;
+}
 
-	if ( cms_chan ) {
+void GameBlaster::WriteDataToLeftDevice(io_port_t, io_val_t value, io_width_t)
+{
+	RenderUpToNow();
+	devices[0]->data_w(0, 0, check_cast<uint8_t>(value));
+}
 
-		//Have there been 10 seconds of no commands, disable channel
-		if ( lastWriteTicks + 10000 < PIC_Ticks ) {
-			cms_chan->Enable( false );
-			return;
-		}
-		int32_t result[BUFFER_SIZE][2];
-		int16_t work[2][BUFFER_SIZE];
-		int16_t* buffers[2] = { work[0], work[1] };
-		device_sound_interface::sound_stream stream;
-		device[0]->sound_stream_update(stream, 0, buffers, len);
-		for (Bitu i = 0; i < len; i++) {
-			result[i][0] = work[0][i];
-			result[i][1] = work[1][i];
-		}
-		device[1]->sound_stream_update(stream, 0, buffers, len);
-		for (Bitu i = 0; i < len; i++) {
-			result[i][0] += work[0][i];
-			result[i][1] += work[1][i];
+void GameBlaster::WriteControlToLeftDevice(io_port_t, io_val_t value, io_width_t)
+{
+	RenderUpToNow();
+	devices[0]->control_w(0, 0, check_cast<uint8_t>(value));
+}
+
+void GameBlaster::WriteDataToRightDevice(io_port_t, io_val_t value, io_width_t)
+{
+	RenderUpToNow();
+	devices[1]->data_w(0, 0, check_cast<uint8_t>(value));
+}
+
+void GameBlaster::WriteControlToRightDevice(io_port_t, io_val_t value, io_width_t)
+{
+	RenderUpToNow();
+	devices[1]->control_w(0, 0, check_cast<uint8_t>(value));
+}
+
+double GameBlaster::ConvertFramesToMs(const int frames)
+{
+	return frames / frame_rate_per_ms;
+}
+
+void GameBlaster::AudioCallback(uint16_t requested_frames)
+{
+	assert(channel);
+	while (requested_frames && fifo.size()) {
+		channel->AddSamples_s16(1, fifo.front().data());
+		fifo.pop();
+		--requested_frames;
+	}
+
+	if (requested_frames) {
+		last_render_time += ConvertFramesToMs(requested_frames);
+		while (requested_frames--) {
+			const auto frame = RenderOnce();
+			channel->AddSamples_s16(1, frame.data());
 		}
-		cms_chan->AddSamples_s32( len, result[0] );
+	}
+	// Pause the card if it hasn't been written to for 10 seconds
+	if (unwritten_for_ms++ > 10000) {
+		is_enabled = false;
+		channel->Enable(false);
+		return;
```

Comment:
Nice optimisation.

Reply:
(:+1: , all credit to original authors - concept was in the original, this is just a small change to make it similar to other Staging implementations).